redditdownloader/processing/controller.py

```diff
@@ -38,9 +38,14 @@ def run(self):
 		[t.join() for t in self._all_processes]
 
 	def stop(self):
+		# Set the stop event for all threads, so there's no new work
 		self.loader.get_stop_event().set()
+		# Give each downloader 2 seconds to finish it's current task (max len(self._downloaders) * 2 seconds)
 		for d in self._downloaders:
-			d.terminate()
+			# if the join times out, whatever is running will be re-downloaded when downloading starts again
+			# since the database is updated on completion
+			d.join(2)
```

Comment:
I like the timeout, but I think we still need to fall back to terminating the threads after waiting, if they are still alive. It won't be good if these keep running in the background, unmanaged.

Reply:
Ahh, I didn't realize it didn't terminate the threads when it timed out.   I've added a check after joining to terminate the thread if it doesn't finish in time.

One thing I'm worried about it the ui hanging when the user has a larger amount of download threads.  I'm considering lowering the timeout value, but checking if the threads terminated after attempting to join them all, but it might look a little messy:

```diff
for d in self._downloaders:
+	d.join(1)
-	d.join(2)
+for d in self._downloaders:
+	if d.is_alive():
+		d.terminate()
```

Reply:
Hmm, that's true. It would be nice to have a hard limit for all threads.

Just off the top of my head, I have something that might work. Maybe the cleanest method would be to set an offset max time using `max_time = time.time()+offset`, then joining each thread for up to `max(0, max_time - time.time() )`, then terminating any that are still running. This would let it exit early if all the threads were ready, but would otherwise wait for a specific window of time - which could be set to a few seconds.

Reply:
It only took almost a year, but I finally made the changes!
Like you reasoned, now it will take the current time and add 5 seconds.
Then it'll loop through threads and attempt to join them until 5 seconds have passed since starting the loop.
Then it'll just force terminate the threads.

From my testing, it should terminate before the 5 second limit most of the time.  
It should just force terminate when doing larger downloads like videos.